core/src/main/java/com/linecorp/armeria/client/limit/ConcurrencyLimit.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.limit;
+
+import static com.linecorp.armeria.client.limit.AbstractConcurrencyLimitingClient.validateMaxConcurrency;
+import static java.lang.Integer.MAX_VALUE;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.RequestContext;
+
+/**
+ * Concurrency settings that limits the concurrent number of active requests.
+ */
+public final class ConcurrencyLimit {
+    private static final long DEFAULT_TIMEOUT_MILLIS = 10000L;
+
+    private final Function<RequestContext, Boolean> policy;
+    private final int maxConcurrency;
+    private final long timeoutMillis;
+    private final AtomicInteger numActiveRequests = new AtomicInteger();
+
+    /**
+     * Creates a new instance to limit the concurrent number of active requests to {@code maxConcurrency}.
+     *
+     * @param policy the policy to apply on the incoming {@code RequestContext} if the request should be
+     *      limited or not.
+     * @param maxConcurrency the maximum number of concurrent active requests. {@code 0} to disable the limit.
+     * @param timeoutMillis the amount of time until this decorator fails the request if the request was not
+     *                delegated to the {@code delegate} before then.
+     */
+    ConcurrencyLimit(Function<RequestContext, Boolean> policy, int maxConcurrency, long timeoutMillis) {
```

Comment:
Could you `Predicate` and `ClientRequestContext` because it is only used for client-side?
```suggestion
    ConcurrencyLimit(Predicate<? super ClientRequestContext> policy, int maxConcurrency, long timeoutMillis) {
```